src/lib/pubkey/sphincsplus/sphincsplus_common/sp_parameters.cpp

```diff
@@ -0,0 +1,206 @@
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ **/
+
+#include <botan/sp_parameters.h>
+
+#include <botan/concepts.h>
+#include <botan/exceptn.h>
+#include <botan/internal/bit_ops.h>
+#include <botan/internal/fmt.h>
+
+#include <cmath>
+
+namespace Botan {
+
+namespace {
+
+Sphincs_Parameter_Set set_from_name(std::string_view name) {
+   if(name == "SphincsPlus-sha2-128s-r3.1" || name == "SphincsPlus-shake-128s-r3.1" ||
+      name == "SphincsPlus-haraka-128s-r3.1")
+      return Sphincs_Parameter_Set::Sphincs128Small;
+   if(name == "SphincsPlus-sha2-128f-r3.1" || name == "SphincsPlus-shake-128f-r3.1" ||
+      name == "SphincsPlus-haraka-128f-r3.1")
+      return Sphincs_Parameter_Set::Sphincs128Fast;
+
+   if(name == "SphincsPlus-sha2-192s-r3.1" || name == "SphincsPlus-shake-192s-r3.1" ||
+      name == "SphincsPlus-haraka-192s-r3.1")
+      return Sphincs_Parameter_Set::Sphincs192Small;
+   if(name == "SphincsPlus-sha2-192f-r3.1" || name == "SphincsPlus-shake-192f-r3.1" ||
+      name == "SphincsPlus-haraka-192f-r3.1")
+      return Sphincs_Parameter_Set::Sphincs192Fast;
+
+   if(name == "SphincsPlus-sha2-256s-r3.1" || name == "SphincsPlus-shake-256s-r3.1" ||
+      name == "SphincsPlus-haraka-256s-r3.1")
+      return Sphincs_Parameter_Set::Sphincs256Small;
+   if(name == "SphincsPlus-sha2-256f-r3.1" || name == "SphincsPlus-shake-256f-r3.1" ||
+      name == "SphincsPlus-haraka-256f-r3.1")
+      return Sphincs_Parameter_Set::Sphincs256Fast;
+
+   throw Lookup_Error(fmt("No SphincsPlus parameter set found for: {}", name));
+}
+
+Sphincs_Hash_Type hash_from_name(std::string_view name) {
+   if(name == "SphincsPlus-sha2-128s-r3.1" || name == "SphincsPlus-sha2-128f-r3.1" ||
+      name == "SphincsPlus-sha2-192s-r3.1" || name == "SphincsPlus-sha2-192f-r3.1" ||
+      name == "SphincsPlus-sha2-256s-r3.1" || name == "SphincsPlus-sha2-256f-r3.1")
+      return Sphincs_Hash_Type::Sha256;
+   if(name == "SphincsPlus-shake-128s-r3.1" || name == "SphincsPlus-shake-128f-r3.1" ||
+      name == "SphincsPlus-shake-192s-r3.1" || name == "SphincsPlus-shake-192f-r3.1" ||
+      name == "SphincsPlus-shake-256s-r3.1" || name == "SphincsPlus-shake-256f-r3.1")
+      return Sphincs_Hash_Type::Shake256;
+   if(name == "SphincsPlus-haraka-128s-r3.1" || name == "SphincsPlus-haraka-128f-r3.1" ||
+      name == "SphincsPlus-haraka-192s-r3.1" || name == "SphincsPlus-haraka-192f-r3.1" ||
+      name == "SphincsPlus-haraka-256s-r3.1" || name == "SphincsPlus-haraka-256f-r3.1")
+      return Sphincs_Hash_Type::Haraka;
+
+   throw Lookup_Error(fmt("No SphincsPlus hash instantiation found for: {}", name));
+}
+
+const char* as_string(Sphincs_Hash_Type type) {
+   switch(type) {
+      case Sphincs_Hash_Type::Sha256:
+         return "sha2";
+      case Sphincs_Hash_Type::Shake256:
+         return "shake";
+      case Sphincs_Hash_Type::Haraka:
+         return "haraka";
+   }
+
+   Botan::unreachable();
```

Comment:
I'd rather just have a `default:` block that has an assertion failure, if we somehow don't have a match, rather than UB.

Reply:
Or returning a `nullptr` or `std::optional<const char*>` if we want this to be `nothrow`